qualtran/bloqs/state_preparation/controlled_state_preparation.py

```diff
@@ -0,0 +1,484 @@
+
+
+_CONTROLLED_STATE_PREP_DOC = BloqDocSpec(
+    bloq_cls=ControlledStatePreparationUsingRotations,
+    import_line='from qualtran.bloqs.state_preparation.controlled_state_preparation import ControlledStatePreparationUsingRotations',
+    examples=(_controlled_state_prep,),
+)
+
+
+@attrs.frozen
+class ControlledQROMRotateQubit(Bloq):
+    r"""Array of controlled rotations $Z^{\theta_i/2}$ for a list of angles $\theta$.
+
+    It uses phase kickback and thus needs a phase gradient state in order to work. This
+    state must be provided externally for efficiency, as it is unaffected and can thus be reused.
+    Refer to [1], section on arbitrary quantum state preparation on page 3.
+
+    Args:
+        selection_bitsizes: number of qubits used for encoding the selection register of the QROM, it
+            must be equal to $\lceil \log_2(l_{rv}) \rceil$, where $l_{rv}$ is the number of angles
+            provided.
+        rot_reg_bitsizes: size of the register that is used to store the rotation angles. Bigger values
+            increase the accuracy of the results.
+        rom_values: the tuple of values to be loaded in the rom, which correspond to the angle of
+            each rotation. In order to get the rom value that corresponds to a given angle use
+            RotationTree.angle_to_rom_value(angle, rot_reg_bitsizes).
+
+    References:
+        [Trading T-gates for dirty qubits in state preparation and unitary synthesis]
+        (https://arxiv.org/abs/1812.00954).
+            Low, Kliuchnikov, Schaeffer. 2018.
+    """
+
+    selection_bitsizes: int
+    rot_reg_bitsizes: int
+    rom_values: Tuple
+
+    @property
+    def signature(self):
+        return Signature.build(
+            prepare_control=1,
+            selection=self.selection_bitsizes,
+            qubit=1,
+            rot_reg=self.rot_reg_bitsizes,
+            phase_gradient=self.rot_reg_bitsizes,
+        )
+
+    def build_composite_bloq(self, bb: BloqBuilder, **soqs: SoquetT) -> Dict[str, SoquetT]:
+        """Parameters:
+        * prepare_control
+        * selection (not necessary if selection_bitsizes == 0)
+        * qubit
+        * rot_reg
+        * phase_gradient
+        """
+        qrom = QROM(
+            [np.array(self.rom_values)],
+            selection_bitsizes=(self.selection_bitsizes,),
+            target_bitsizes=(self.rot_reg_bitsizes,),
+            num_controls=2,
+        )
+        # both prepare_control and qubit will control the QROM so that this acts as a control Z
+        qrom_control = bb.join(np.array([soqs["prepare_control"], soqs["qubit"]]))
+        # load angles in rot_reg (line 1 of eq (8) in [1])
+        qrom_control, soqs = self._apply_qrom(qrom, bb, qrom_control, soqs)
+        # phase kickback via phase_grad += rot_reg (line 2 of eq (8) in [1])
+        soqs["rot_reg"], soqs["phase_gradient"] = bb.add(
+            Add(bitsize=self.rot_reg_bitsizes), a=soqs["rot_reg"], b=soqs["phase_gradient"]
+        )
```

Comment:
We also have a `AddIntoPhaseGrad` gate which has a slightly better Toffoli complexity and can be used here after slight modifications. 

Add the following function to `AddIntoPhaseGrad` 

```python
    def add_my_tensors(
        self,
        tn: 'qtn.TensorNetwork',
        tag: Any,
        *,
        incoming: Dict[str, 'SoquetT'],
        outgoing: Dict[str, 'SoquetT'],
    ):
        import quimb.tensor as qtn

        N, M = 2**self.inp_bitsize, 2**self.phase_bitsize
        inds = (incoming['x'], incoming['phase_grad'], outgoing['x'], outgoing['phase_grad'])
        unitary = np.zeros((N, M) * 2, dtype=np.complex128)
        for a, b in itertools.product(range(N), range(M)):
            unitary[a, b, a, int(math.fmod(a + b, M))] = 1

        tn.add(qtn.Tensor(data=unitary, inds=inds, tags=[self.short_name(), tag]))
```

and modify this code as 

```python
        soqs["rot_reg"], soqs["phase_gradient"] = bb.add(
            AddIntoPhaseGrad(self.rot_reg_bitsizes, self.rot_reg_bitsizes),
            x=soqs["rot_reg"],
            phase_grad=soqs["phase_gradient"],
        )
```